invariant/analyzer/language/parser.py

```diff
@@ -31,39 +32,40 @@
     decl_stmt: ( ID | func_signature ) ":=" expr | ( ID | func_signature ) INDENT NEWLINE? expr (NEWLINE expr)* DEDENT
 
     expr: ID | assignment_expr | "(" expr ("," expr)* ")" | block_expr | import_stmt
-    
+
     quantifier_expr: ( "not" )? ( func_call | ID ) INDENT NEWLINE? expr (NEWLINE expr)* DEDENT
     block_expr: INDENT expr (NEWLINE expr)* DEDENT
-    
+
     assignment_expr: ( ID ":=" binary_expr ) | binary_expr
     binary_expr: cmp_expr LOGICAL_OPERATOR cmp_expr | cmp_expr
     cmp_expr: ( term CMP_OPERATOR term ) | term
     term: factor TERM_OPERATOR factor | factor
     factor: power FACTOR_OPERATOR power | power
     power: atom POWER_OPERATOR atom | atom
-    atom: unary_expr | NUMBER | multiline_string | STRING | ID | "(" expr ")" | member_access | key_access | expr | func_call | quantifier_expr | typed_identifier | tool_ref | object_literal | list_literal | STAR | value_ref
-                   
+    atom: unary_expr | NUMBER | multiline_string | STRING | ID | "(" expr ")" | member_access | key_access | expr | func_call | quantifier_expr | typed_identifier | tool_ref | object_literal | list_literal | STAR | value_ref | list_comprehension | ternary_op
+
     unary_expr: UNARY_OPERATOR expr
     func_call: expr  "(" ( (expr ("," expr)*)? ("," kwarg ("," kwarg)*)? ) ")" | \
                expr  "(" (kwarg ("," kwarg)*)? ")"
     kwarg: ID "=" expr
     func_signature: ID "(" (parameter_decl ("," parameter_decl)*)? ")"
     parameter_decl: ID ":" ID
-    
+
     member_access: expr "." ID
     key_access: expr "[" expr "]"
 
     typed_identifier: "(" ID ":" ID ")"
     tool_ref: "tool" ":" ID
-    
+
     object_literal: "{" ( object_entry ("," object_entry)* )? "}"
     object_entry: (ID|STRING) ":" expr
     list_literal: "[" ( expr ("," expr)* )? "]"
+    list_comprehension: "[" expr "for" ID "in" expr ("if" expr)? "]"
```

Comment:
note: with this, we don't support destructuring enumerations right now e.g. `[a for (a,b) in enumerate(something)]` right now (which is fine ofc).